bpe/apply.py

```diff
@@ -95,3 +95,45 @@ def run(self):
         tree = ET.ElementTree(elem)
         with gzip.open(self.converted_lexicon.get_path(), "wb") as f:
             tree.write(f, encoding="utf-8")
+
+
+class ApplyBPEToTextJob(Job):
+    """
+    Apply BPE codes on a text file
+    """
+
+    def __init__(self, text_file, bpe_codes, bpe_vocab=None):
+        """
+        :param Path|str text_file: words text file to convert to bpe
+        :param Path|str bpe_codes: bpe codes file
+        :param Path|str|None: bpe_vocab: if provided, then merge operations that produce OOV are reverted
+        """
+        self.text_file = text_file
+        self.bpe_codes = bpe_codes
+        self.bpe_vocab = bpe_vocab
+
+        self.subword_nmt_repo = CloneGitRepositoryJob(
```

Comment:
I think it's better to have this as an argument to the job. It makes it a bit more complicated to use the job, but you gain more flexibility. Eventually there might be changes in the NMT repo that make you want to update to a different version.

Reply:
We agreed that this won't change. That's why I also use a fixed commit hash. This anw might get deprecated later as we plan in the future to use `sentencepiece` software which can generates BPEs, Sentencepices, etc. This is more generic and already support by RETURNN.

Reply:
Still this should definitely be outside of the job

Reply:
hmm ok. this will make it a bit complicated but ok i can just pass the repo as an argument.

Reply:
Shouldn't we handle this like we handle other tools or external dependencies like RETURNN, RASR? In the usual case, I would not want to explicitly pass this as an argument, but instead just install it somewhere and then Sisyphus would pick my default.

Reply:
Does a switch to `sentencepiece` mean you want to extend the job, deprecate subword_nmt or create new jobs?
I am not sure if we want to deprecate subword_nmt in a year or so, since then people will be relying on these jobs..

> Shouldn't we handle this like we handle other tools or external dependencies like RETURNN, RASR?

the default can be specified in `settings.py` but you can also pass a specific path to the job

Reply:
> Does a switch to `sentencepiece` mean you want to extend the job, deprecate subword_nmt or create new jobs?

The job cannot really be extended/changed when `subword_nmt` is an argument, right?

SentencePiece also creates other types of files. So it probably doesn't make sense to use the same job.

> I am not sure if we want to deprecate subword_nmt in a year or so, since then people will be relying on these jobs..
> 
> > Shouldn't we handle this like we handle other tools or external dependencies like RETURNN, RASR?
>
> the default can be specified in `settings.py` but you can also pass a specific path to the job

You mean for Subword-NMT and SentencePiece?

Reply:
> Does a switch to sentencepiece mean you want to extend the job, deprecate subword_nmt or create new jobs?
I am not sure if we want to deprecate subword_nmt in a year or so, since then people will be relying on these jobs..
No this would be new job. this will not affect this job so the discussion is independent to this PR.

Reply:
ok so how should we proceed here? I can define the subword-nmt repo in `settings.py` and then use it in the job as `gs.SUBWORD_NMT` or so. but also i can keep the repo as parameter and set it to default to None. So if this was `None`, it will just use the one defined in `settings.py`

Reply:
Yes, I would do it exactly like the RETURNN jobs do it